pkg/storage/endpoint/gc_states.go

```diff
@@ -0,0 +1,556 @@
+	prefix := keypath.CompatibleTiDBMinStartTSPrefix()
+	if keyspaceID != constant.NullKeyspaceID {
+		prefix = keypath.CompatibleKeyspaceTiDBMinStartTSPrefix(keyspaceID)
+	}
+	prefixEnd := clientv3.GetPrefixRangeEnd(prefix)
+
+	// TODO: Limit the count for each call.
+	keys, values, err := p.storage.LoadRange(prefix, prefixEnd, 0)
+	if err != nil {
+		return "", 0, err
+	}
+
+	if len(keys) == 0 {
+		return "", 0, nil
+	}
+
+	var minKey string
+	var minMinStartTS uint64
+
+	for i, valueStr := range values {
+		minStartTS, err := strconv.ParseUint(valueStr, 10, 64)
+		if err != nil {
+			return "", 0, errs.ErrStrconvParseUint.Wrap(err).GenWithStackByArgs()
+		}
+		if len(minKey) == 0 || minStartTS < minMinStartTS {
+			minMinStartTS = minStartTS
+			minKey = keys[i]
+		}
+	}
+
+	// Remove prefix from the key and only keep the identifier written by TiDB.
+	if len(minKey) < len(prefix) || minKey[:len(prefix)] != prefix {
+		// This is expected to be unreachable.
+		return "", 0, errors.Errorf("unexpected internal error: loading TiDB min start ts but got mismatching key prefix, expected prefix: %s, got key: %s", prefix, minKey)
+	}
+	minKey = minKey[len(prefix):]
+	return minKey, minMinStartTS, nil
+}
+
+// RunInGCStateTransaction runs a transaction for updating GC states or read a batch of GC states.
+// The atomicity is guaranteed by a "revision" key. Any non-empty write caused by the transaction will increase the
+// revision.
+// In the transaction, reads can be performed on the GCStateProvider as usual, while writes should only be performed
+// through the GCStateWriteBatch.
+func (p GCStateProvider) RunInGCStateTransaction(f func(wb *GCStateWriteBatch) error) error {
+	revisionKey := keypath.GCStateRevisionPath()
+	currentRevision, err := p.storage.Load(revisionKey)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+	condition := kv.RawTxnCondition{
+		Key:     revisionKey,
+		CmpType: kv.RawTxnCmpNotExists,
+	}
+	var currentRevisionValue uint64
+	if currentRevision != "" {
+		condition.CmpType = kv.RawTxnCmpEqual
+		condition.Value = currentRevision
+		currentRevisionValue, err = strconv.ParseUint(currentRevision, 10, 64)
+	}
+
+	if err != nil {
+		return errors.AddStack(err)
+	}
+	nextRevision := fmt.Sprintf("%d", currentRevisionValue+1)
+
+	wb := GCStateWriteBatch{}
+	err = f(&wb)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	ops := wb.ops
+
+	// No need to increase the revision if there's no write (so that it acts like an RLock and concurrent reads won't
+	// conflict with each other).
+	if len(ops) > 0 {
+		ops = append(ops, kv.RawTxnOp{
+			Key:    revisionKey,
+			OpType: kv.RawTxnOpPut,
+			Value:  nextRevision,
+		})
+	}
+
+	txn := p.storage.CreateRawTxn()
+	result, err := txn.If(condition).Then(ops...).Commit()
+	if err != nil {
+		return errors.AddStack(err)
+	}
+	if !result.Succeeded {
+		return errs.ErrEtcdTxnConflict.GenWithStackByArgs()
+	}
+
+	if len(ops) != len(result.Responses) {
```

Comment:
If this happens, do we need to rollback or manually fix it?

Reply:
If I understanding is correct, etcd shoudl guarantee this to be true. I expect that this path is never reachable. If this happen, I think it should be a bug in etcd or etcd-client and it doesn't look possible to automatically fix it in a simple way. And as the `Commit` returns no error here, it seems we no longer have chance to roll it back atomically. I think if this does happen, we may have no choice but to manually recover it.